WcaOnRails/app/models/competition.rb

```diff
@@ -1117,7 +1119,11 @@ def has_date_errors?
   end
 
   # The competition must be at least 28 days in advance in order to confirm it. Admins are able to modify the competition despite being less than 28 days in advance.
-  validate :start_date_must_be_28_days_in_advance, if: :confirmed_or_visible?
+  # We only run this validation if we're actually changing the start_date or
+  # confirming the competition, to not prevent organizers/delegates from
+  # updating competition-specific setttings, such as the receive notifications checkbox.
+  validate :start_date_must_be_28_days_in_advance,
+           if: -> { confirmed_or_visible? && (will_save_change_to_start_date? || will_save_change_to_confirmed_at?) }
```

Comment:
Can we move this out into its own method, along the lines of `changes_critical_dates?`?

Reply:
To be honest I don't quite see the advantage of that *if that method isn't reused* by some other part of the code; it also feels quite common to me to use a lambda in this circumstance.

If you feel strongly about this I can definitely change that part!

Reply:
It's not necessarily about code reuse. I do feel rather strongly about this topic simply because I don't want too much to happen within one (or two) lines of code.

The primary purpose of the code sections with the validations should be (imo) to declare _that validations exist_. The specifics of _how_ to validate things, and _when_, is a separate concern from a high-level point of view. This is especially true when the validation has some higher semantic meaning within our business logic (as is the case here: Does the user change critical details of the competition?). Extracting it does not only help with reusing the method (to which I agree that it likely won't happen in this specifc case) but also to give a concise name to the lambda by virtue of declaring it as a full-fledged method, so that other contributors can figure out what's going on at a glance and without reading several lines of comments.

TL;DR: The `if ->` lambda you're constructing here has a meaning, and I believe we should express that meaning by moving it into a function with a name.

(Sidenote: I know that the strict consequence of this would be a huge restructure of the entire `competition.rb` file, but hey, you gotta start somewhere, right? 😸)

Reply:
Change done and PR rebased!

I would agree to outline the function if it wasn't a one-liner or if it was re-used somewhere else, but for one-liners a lambda looks perfectly fine to me (the `start_date_must_be_28_days_in_advance` is actually a very good example of what should be in its own method IMO!).

> but also to give a concise name to the lambda by virtue of declaring it as a full-fledged method, so that other contributors can figure out what's going on at a glance and without reading several lines of comments.

By this logic, what's the point of commenting at all? Comments are meant to be read: the code should be clear on what it does, but why/how this or that needs to be done seems perfectly appropriate for a comment.

> (Sidenote: I know that the strict consequence of this would be a huge restructure of the entire competition.rb file, but hey, you gotta start somewhere, right? smile_cat)

A strict consequence of this would be a huge restructure of the competition.rb file, pretty much all our models, the [rails guide about validations](https://guides.rubyonrails.org/active_record_validations.html#using-a-proc-with-if-and-unless) which explicitly recommends lambda for one-liners, and pretty much all code bases which follow this guide :smiley:

Reply:
Thanks! I appreciate that this is a subjective discussion with different opinions, and I most certainly don't want to overthrow the Rails guide single-handedly. However, the complexity of what a "one-liner" (quote from that guide) actually means is difficult to define.

> By this logic, what's the point of commenting at all? Comments are meant to be read: the code should be clear on what it does, but why/how this or that needs to be done seems perfectly appropriate for a comment.

I absolutely agree. Which is why in this case, with clear and concise method names, the first half of the comment seems a bit redundant. The idea should be that comments are there to explain stuff that is too complex to "explain" in the code (by virtue of giving good names to variables / functions in the first place) directly.

Also, you opted for the rather simple choice of calling the extracted method `should_validate_start_date?`. That's technically not wrong, but only because right now it is exclusively used as an if-check in the Rails model validation. My intention behind the original request was that this method seems so useful that it could potentially be used in other places in the future, along the lines of `is_chaging_critical_data?`. That way, having it be a standalone method makes a lot more sense and I agree that the current naming scheme feels a bit overkill.

Anyways, I don't want to hold up this feature by more philosophy deliberations. Merging ASAP :D

Reply:
Oh about the naming: your initial suggestion was regarding *dates*, which is why I didn't keep it in the first place; I did try to find a more describing name but failed to find something real nice (`is_changing_critical_data?` isn't quite good either because what is "critical" is very dependent on the context); maybe something like `is_confirming_competition?` could have been a good candidate?

Reply:
Yes, that sounds like a good option! Or maybe `is_changing_announcement_dates?`?